From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: eins0fx <eins0fx@users.noreply.github.com>
Date: Mon, 16 Dec 2024 00:00:00 +0700
Subject: [PATCH] HID: magicmouse: Add T2 trackpad support for kernel 6.18+

This patch adds support for trackpads found on Macs with the T2
Security Chip for kernel 6.18+. The upstream hid-magicmouse.c was
significantly refactored, so this is a rewrite of the T2 support.

Supported devices:
- MacBook Pro 13" 2018-2019 (J140K)
- MacBook Pro 15" 2018-2019 (J132)
- MacBook Pro 16" 2019 (J152F, J680)
- MacBook Air 2018-2020 (J213, J214K, J223, J230K)

Signed-off-by: eins0fx <eins0fx@users.noreply.github.com>
---
--- a/drivers/hid/hid-magicmouse.c	2025-12-13 01:42:47.000000000 +0800
+++ b/drivers/hid/hid-magicmouse.c	2025-12-16 00:45:00.223254711 +0800
@@ -112,6 +112,42 @@
 #define TRACKPAD2_RES_Y \
 	((TRACKPAD2_MAX_Y - TRACKPAD2_MIN_Y) / (TRACKPAD2_DIMENSION_Y / 100))
 
+/* T2 MacBook trackpad dimensions - MacBook Pro 13" 2018-2019 */
+#define J140K_TP_DIMENSION_X (float)12100
+#define J140K_TP_MIN_X -5318
+#define J140K_TP_MAX_X 5787
+#define J140K_TP_RES_X \
+	((J140K_TP_MAX_X - J140K_TP_MIN_X) / (J140K_TP_DIMENSION_X / 100))
+#define J140K_TP_DIMENSION_Y (float)8200
+#define J140K_TP_MIN_Y -157
+#define J140K_TP_MAX_Y 7102
+#define J140K_TP_RES_Y \
+	((J140K_TP_MAX_Y - J140K_TP_MIN_Y) / (J140K_TP_DIMENSION_Y / 100))
+
+/* MacBook Pro 15" 2018-2019 */
+#define J132_TP_DIMENSION_X (float)13500
+#define J132_TP_MIN_X -6243
+#define J132_TP_MAX_X 6749
+#define J132_TP_RES_X \
+	((J132_TP_MAX_X - J132_TP_MIN_X) / (J132_TP_DIMENSION_X / 100))
+#define J132_TP_DIMENSION_Y (float)8400
+#define J132_TP_MIN_Y -170
+#define J132_TP_MAX_Y 7685
+#define J132_TP_RES_Y \
+	((J132_TP_MAX_Y - J132_TP_MIN_Y) / (J132_TP_DIMENSION_Y / 100))
+
+/* MacBook Pro 16" 2019 (J152F/J680) */
+#define J152F_TP_DIMENSION_X (float)16000
+#define J152F_TP_MIN_X -7456
+#define J152F_TP_MAX_X 7976
+#define J152F_TP_RES_X \
+	((J152F_TP_MAX_X - J152F_TP_MIN_X) / (J152F_TP_DIMENSION_X / 100))
+#define J152F_TP_DIMENSION_Y (float)10200
+#define J152F_TP_MIN_Y -163
+#define J152F_TP_MAX_Y 9283
+#define J152F_TP_RES_Y \
+	((J152F_TP_MAX_Y - J152F_TP_MIN_Y) / (J152F_TP_DIMENSION_Y / 100))
+
 /**
  * struct magicmouse_sc - Tracks Magic Mouse-specific data.
  * @input: Input device through which we report events.
@@ -151,6 +187,24 @@
 	struct timer_list battery_timer;
 };
 
+static bool magicmouse_is_t2_trackpad(struct hid_device *hdev)
+{
+	switch (hdev->product) {
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J140K:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J132:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J680:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J680_ALT:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J213:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J214K:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J223:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J230K:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J152F:
+		return true;
+	default:
+		return false;
+	}
+}
+
 static int magicmouse_firm_touch(struct magicmouse_sc *msc)
 {
 	int touch = -1;
@@ -231,7 +285,8 @@
 		down = state != TOUCH_STATE_NONE;
 	} else if (input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD2 ||
 		   input->id.product ==
-			   USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC) {
+			   USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC ||
+	   magicmouse_is_t2_trackpad(msc->hdev)) {
 		id = tdata[8] & 0xf;
 		x = (tdata[1] << 27 | tdata[0] << 19) >> 19;
 		y = -((tdata[3] << 30 | tdata[2] << 22 | tdata[1] << 14) >> 19);
@@ -265,7 +320,8 @@
 	 */
 	if (emulate_scroll_wheel &&
 	    input->id.product != USB_DEVICE_ID_APPLE_MAGICTRACKPAD2 &&
-	    input->id.product != USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC) {
+	    input->id.product != USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC &&
+	    !magicmouse_is_t2_trackpad(msc->hdev)) {
 		unsigned long now = jiffies;
 		int step_x = msc->touches[id].scroll_x - x;
 		int step_y = msc->touches[id].scroll_y - y;
@@ -366,7 +422,8 @@
 
 		if (input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD2 ||
 		    input->id.product ==
-			    USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC)
+			    USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC ||
+		    magicmouse_is_t2_trackpad(msc->hdev))
 			input_report_abs(input, ABS_MT_PRESSURE, pressure);
 
 		if (report_undeciphered) {
@@ -506,7 +563,8 @@
 		input_report_rel(input, REL_Y, y);
 	} else if (input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD2 ||
 		   input->id.product ==
-			   USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC) {
+			   USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC ||
+		   magicmouse_is_t2_trackpad(hdev)) {
 		input_mt_sync_frame(input);
 		input_report_key(input, BTN_MOUSE, clicks & 1);
 	} else { /* USB_DEVICE_ID_APPLE_MAGICTRACKPAD */
@@ -593,6 +651,18 @@
 
 		mt_flags = INPUT_MT_POINTER | INPUT_MT_DROP_UNUSED |
 				INPUT_MT_TRACK;
+	} else if (magicmouse_is_t2_trackpad(hdev)) {
+		/* T2 MacBook Trackpad - similar to Magic Trackpad 2 */
+		__clear_bit(EV_MSC, input->evbit);
+		__clear_bit(BTN_0, input->keybit);
+		__clear_bit(BTN_RIGHT, input->keybit);
+		__clear_bit(BTN_MIDDLE, input->keybit);
+		__set_bit(BTN_MOUSE, input->keybit);
+		__set_bit(INPUT_PROP_BUTTONPAD, input->propbit);
+		__set_bit(BTN_TOOL_FINGER, input->keybit);
+
+		mt_flags = INPUT_MT_POINTER | INPUT_MT_DROP_UNUSED |
+				INPUT_MT_TRACK;
 	} else { /* USB_DEVICE_ID_APPLE_MAGICTRACKPAD */
 		/* input->keybit is initialized with incorrect button info
 		 * for Magic Trackpad. There really is only one physical
@@ -661,6 +731,24 @@
 		input_abs_set_res(input, ABS_Y, TRACKPAD2_RES_Y);
 		input_abs_set_res(input, ABS_MT_POSITION_X, TRACKPAD2_RES_X);
 		input_abs_set_res(input, ABS_MT_POSITION_Y, TRACKPAD2_RES_Y);
+	} else if (magicmouse_is_t2_trackpad(hdev)) {
+		/* T2 Trackpad - use J152F dimensions (16" MacBook Pro) */
+		input_set_abs_params(input, ABS_MT_PRESSURE, 0, 253, 0, 0);
+		input_set_abs_params(input, ABS_PRESSURE, 0, 253, 0, 0);
+		input_set_abs_params(input, ABS_MT_ORIENTATION, -3, 4, 0, 0);
+		input_set_abs_params(input, ABS_X, J152F_TP_MIN_X,
+				     J152F_TP_MAX_X, 0, 0);
+		input_set_abs_params(input, ABS_Y, J152F_TP_MIN_Y,
+				     J152F_TP_MAX_Y, 0, 0);
+		input_set_abs_params(input, ABS_MT_POSITION_X,
+				     J152F_TP_MIN_X, J152F_TP_MAX_X, 0, 0);
+		input_set_abs_params(input, ABS_MT_POSITION_Y,
+				     J152F_TP_MIN_Y, J152F_TP_MAX_Y, 0, 0);
+
+		input_abs_set_res(input, ABS_X, J152F_TP_RES_X);
+		input_abs_set_res(input, ABS_Y, J152F_TP_RES_Y);
+		input_abs_set_res(input, ABS_MT_POSITION_X, J152F_TP_RES_X);
+		input_abs_set_res(input, ABS_MT_POSITION_Y, J152F_TP_RES_Y);
 	} else { /* USB_DEVICE_ID_APPLE_MAGICTRACKPAD */
 		input_set_abs_params(input, ABS_MT_ORIENTATION, -31, 32, 1, 0);
 		input_set_abs_params(input, ABS_X, TRACKPAD_MIN_X,
@@ -711,7 +799,8 @@
 	if ((hi->input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD ||
 	     hi->input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD2 ||
 	     hi->input->id.product ==
-		     USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC) &&
+		     USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC ||
+	     magicmouse_is_t2_trackpad(hdev)) &&
 	    field->flags & HID_MAIN_ITEM_RELATIVE)
 		return -1;
 
@@ -1023,6 +1112,24 @@
 		USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC), .driver_data = 0 },
 	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
 		USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J140K), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J132), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J680), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J680_ALT), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J213), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J214K), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J223), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J230K), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J152F), .driver_data = 0 },
 	{ }
 };
 MODULE_DEVICE_TABLE(hid, magic_mice);
-- 
2.47.0
